src/main/java/com/iota/iri/network/pipeline/ReceivedStage.java

```diff
@@ -46,6 +50,15 @@ public ProcessingContext process(ProcessingContext ctx) {
         Neighbor originNeighbor = payload.getOriginNeighbor();
         TransactionViewModel tvm = payload.getTransactionViewModel();
 
+        // free up the recently requested transaction set
+        if(transactionRequester.wasTransactionRecentlyRequested(tvm.getHash())){
+            transactionRequester.removeRecentlyRequestedTransaction(tvm.getHash());
+            // as the transaction came from the request queue, we can add its branch and trunk to the request
+            // queue already, as we only have transactions in the request queue which are needed for solidifying
+            // milestones. this speeds up solidification significantly
+            transactionRequester.requestTrunkAndBranch(tvm);
```

Comment:
>If it is not in the cache for some reason then we will go through all the stages. This is wasteful because when updateStatus(tvm) is invoked we currently check whether we can quickly solidify the tvm. If the answer is "yes" obviously we don't have to request anything.

Can you elaborate why this is wasteful?

Reply:
>Even this change doesn't take care of all the edge cases. Because when you do updateStatus you may not get solid even if the trunk and branch are in the db.  

Why should the node be able to set the newly received transaction to be solid when it was requested for solidifying a milestone?

Reply:
>So I will say that maybe this logic should be in the Reply stage in order to fix case 2a.
This will require changing the payLoad :-(.
However, it is better to create changes that don't introduce any regressions (at least the ones we notice).

What logic should be in the Reply stage?

We should not be removing txs which were cached from the recently seen transactions set because we can not identify whether the same transaction is still in-flight in the ReceivedStage. Doing so may prevent that the transaction's trunk/branch are requested, because the PreProcess/Reply stages removed the tx from the recently requested list.

Reply:
I suggest following: Lets simply clear the recently requested transaction set once the node becomes solid?